test/kpro_fetch_tests.erl

```diff
@@ -81,21 +88,45 @@ drop_older_offsets(Offset, [#kafka_message{offset = O} | R] = ML) ->
     false -> ML
   end.
 
-validate_messages([], Rest) -> Rest;
-validate_messages([#kafka_message{key = K, value = V} | R1], [Msg | R2]) ->
-  ok = validate_message(K, V, Msg),
-  validate_messages(R1, R2).
+validate_messages(_Topic, _FetchVsn, [], Rest) -> Rest;
+validate_messages(Topic, FetchVsn, [#kafka_message{ts_type = TType, ts = T, key = K, value = V} | R1],
+                  [{ProduceVsn, LAT, Msg} | R2]) ->
+  Magic = message_magic(FetchVsn, ProduceVsn),
+  ok =
+    case ?TOPIC_LAT == Topic of
+      true -> validate_message_append(Magic, TType, T, K, V, LAT, Msg);
+      false -> validate_message_create(Magic, TType, T, K, V, Msg)
+    end,
+  validate_messages(Topic, FetchVsn, R1, R2).
 
-validate_message(K, V, {K, V}) -> ok;
-validate_message(K, V, {_T, K, V}) -> ok;
-validate_message(K, V, #{key := K, value := V}) -> ok;
-validate_message(K, V, Wat) ->
-  erlang:error(#{ fetched => {K, V}
+validate_message_create(_Magic = 0, _TType, _T, K, V, #{key := K, value := V}) -> ok;
+validate_message_create(_Magic, create, T, K, V, #{ts := T, key := K, value := V}) -> ok;
+validate_message_create(Magic, TType, T, K, V, Wat) ->
+  erlang:error(#{ fetched => {T, K, V}
                 , produced => Wat
+                , magic => Magic
+                , ts_type => TType
                 }).
 
-do_fetch(Connection, Vsn, BeginOffset, MaxBytes) ->
-  Req = make_req(Vsn, BeginOffset, MaxBytes),
+validate_message_append(_Magic = 0, _TType, _T, K, V, _LAT, #{key := K, value := V}) -> ok;
```

Comment:
This check is getting pretty hard to understand with all these new parameters. Suggestion: pack all the test and reference data into maps, and match only fields you need?